Turns out we'd baked some expectations of [this into tests](https://github.com/moby/moby/actions/runs/11052960324/job/30706287545?pr=48552#step:7:97):
```console
tests/integration/models_containers_test.py:363: in test_exec_run_failed
    assert exec_output[0] == 126
E   assert 127 == 126
```

https://github.com/docker/docker-py/blob/a3652028b1ead708bd9191efb286f909ba6c2a49/tests/integration/models_containers_test.py#L356-L363

Which leads me to wonder if we can reasonably fix this without breaking a lot of users. I want to believe that it's quite a specific edge case (`126`vs`127` when the executable can't be found), but 😨

--- a/daemon/exec.go
+++ b/daemon/exec.go
@@ -183,8 +183,12 @@ func (daemon *Daemon) ContainerExecStart(ctx context.Context, name string, optio
 			ec.Lock()
 			ec.Container.ExecCommands.Delete(ec.ID)
 			ec.Running = false
-			exitCode := 126
-			ec.ExitCode = &exitCode
+			if ec.ExitCode == nil {
+				// default to `126` (`EACCES`) if we fail to start
+				// the exec without setting an exit code.
+				exitCode := exitEaccess
+				ec.ExitCode = &exitCode
+			}
 			if err := ec.CloseStreams(); err != nil {
 				log.G(ctx).Errorf("failed to cleanup exec %s streams: %s", ec.Container.ID, err)
 			}

--- a/hack/make/test-docker-py
+++ b/hack/make/test-docker-py
@@ -25,6 +25,9 @@ fi
 # TODO(vvoland): re-enable after https://github.com/docker/docker-py/pull/3203 is included in the DOCKER_PY_COMMIT release.
 PY_TEST_OPTIONS="$PY_TEST_OPTIONS --deselect=tests/integration/api_image_test.py::CommitTest::test_commit"
 PY_TEST_OPTIONS="$PY_TEST_OPTIONS --deselect=tests/integration/api_image_test.py::CommitTest::test_commit_with_changes"
+
+# TODO(laurazard): re-enable after https://github.com/docker/docker-py/pull/3290 is included in the DOCKER_PY_COMMIT release.
+PY_TEST_OPTIONS="$PY_TEST_OPTIONS --deselect=tests/integration/models_containers_test.py::ContainerTest::test_exec_run_failed"
 (
 	bundle .integration-daemon-start
 

--- a/integration/container/exec_linux_test.go
+++ b/integration/container/exec_linux_test.go
@@ -30,3 +30,36 @@ func TestExecConsoleSize(t *testing.T) {
 	assert.NilError(t, err)
 	assert.Equal(t, strings.TrimSpace(result.Stdout()), "57 123")
 }
+
+func TestFailedExecExitCode(t *testing.T) {
+	testCases := []struct {
+		doc              string
+		command          []string
+		expectedExitCode int
+	}{
+		{
+			doc:              "executable not found",
+			command:          []string{"nonexistent"},
+			expectedExitCode: 127,
+		},
+		{
+			doc:              "executable cannot be invoked",
+			command:          []string{"/etc"},
+			expectedExitCode: 126,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.doc, func(t *testing.T) {
+			ctx := setupTest(t)
+			apiClient := testEnv.APIClient()
+
+			cID := container.Run(ctx, t, apiClient)
+
+			result, err := container.Exec(ctx, apiClient, cID, tc.command)
+			assert.NilError(t, err)
+
+			assert.Equal(t, result.ExitCode, tc.expectedExitCode)
+		})
+	}
+}

